cupy/cuda/cub.pyx

```diff
@@ -96,8 +111,15 @@ def device_reduce(ndarray x, op, out=None, bint keepdims=False):
     cdef void *y_ptr
     cdef void *ws_ptr
     cdef Stream_t s
+    cdef tuple out_shape
+
+    if keepdims:
+        out_shape = tuple([x.shape[axis] if axis in out_axis else 1
+                           for axis in range(x.ndim)])
+        ndim_out = len(out_shape)
+    else:
+        ndim_out = 0
```

Comment:
Can't we just call `_preprocess_array()` here and be more symmetric w.r.t. `device_segmented_reduce`? I know it has a few more extra lines, but the overhead should be moderate?

Reply:
This was also refactored in f8a3db2. I just moved the out_shape computation into a separate helper and call that same function from both `device_reduce` and `device_segmented_reduce`.